pytorch_lightning/utilities/profiler.py

```diff
@@ -0,0 +1,178 @@
+"""
+Profiling your training run can help you understand if there are any bottlenecks in your code.
+
+PyTorch Lightning supports profiling standard actions in the training loop out of the box, including:
+- on_epoch_start
+- on_epoch_end
+- on_batch_start
+- tbptt_split_batch
+- model_forward
+- model_backward
+- on_after_backward
+- optimizer_step
+- on_batch_end
+- training_end
+- on_training_end
+
+If you only wish to profile the standard actions, you can construct a Profiler object and simply
+pass it into the Trainer.
+
+.. code-block:: python
+    profiler = Profiler()
+    trainer = Trainer(..., profiler=profiler)
+
+You can also reference this profiler to profiler any arbitrary code.
+
+.. code-block:: python
+    with profiler.profile('my_custom_action'):
+        my_custom_action()
+"""
+
+
+from contextlib import contextmanager
+from collections import defaultdict
+import time
+import numpy as np
+import cProfile
+import pstats
+import io
+from abc import ABC, abstractmethod
+
+
+class BaseProfiler(ABC):
+    @abstractmethod
+    def start(self, action_name):
+        """
+        defines how to start recording an action
+        """
+        pass
+
+    @abstractmethod
+    def stop(self, action_name):
+        """
+        defines how to record the duration once an action is complete
+        """
+        pass
+
+    @contextmanager
+    def profile(self, action_name):
+        """
+        yields a context manager to encapsulate the scope of a profiled action
+
+        with self.profile('load training data'):
+            # load training data code
+
+        the profiler will start once you've entered the context and automatically stop
+        once you exit the code block
+        """
+        try:
+            self.start(action_name)
+            yield action_name
+        finally:
+            self.stop(action_name)
+
+    def describe(self):
+        """
+        prints a report after the conclusion of the profiled training run
+        """
+        pass
+
+
+class PassThroughProfiler(BaseProfiler):
+    """
+    this can be used when you don't want to profile your runs
+    """
+
+    def __init__(self):
+        pass
+
+    def start(self, action_name):
+        pass
+
+    def stop(self, action_name):
+        pass
+
+
+class Profiler(BaseProfiler):
+    """
+    this profiler simply records the duration of actions (in seconds) and reports
+    the mean and standard deviation of each action duration over the entire training run
+    """
+
+    def __init__(self):
+        self.current_actions = {}
+        self.recorded_durations = defaultdict(list)
+
+    def start(self, action_name):
+        if action_name in self.current_actions:
+            raise ValueError(
```

Comment:
I would consider some kind of soft information (just a logger warning) since it is a complementary service...

Reply:
I gave it some thought and would prefer to keep the exception. For the default profiled actions, this will be tested and we'll ensure that everything works properly (verifying that no tests raise the exception). If the user is profiling additional actions, I would think they would want to know when their code is incorrect. The only time this error is raised is if someone calls `profiler.stop()` without calling `profiler.start()` or vice versa. 

In reality, I don't even think we should document the `start()` and `stop()` interfaces to the user (since it can quickly get very messy), following the principle of having one way to do things. 

> **Force User Decisions To Best Practices**
> There are 1,000 ways to do something. However, something eventually becomes standard practice that everyone does. Thus we pick one way of doing it and force everyone to do it this way.